### PR TITLE
Hostfrequencyscaling

### PR DESCRIPTION
**Pull request by ghazanfarttu**
This module performs mutations related to scaling of CPU frequency to control CPU thermal conditions of HPC node using in-band mechanism.

This module clocks HPC node to "schedutil" scaling governor on boot and whenever CPU temperature reaches to high (warning) condition, module mutates the scaling governor to "powersave".  Current implementation handles critical CPU temperature same as high CPU temperature.

Additionally, there are many other mutations intended for different use cases (e.g. switching back to "schedutil" after "powersave") are under considerations and investigations.

### CODE REVIEW

File: modules/hostfrequencyscaling/hostfrequencyscaling.go [hunk 1]
```diff
@@ -0,0 +1,516 @@
...
+}
+
+// modify this if you want excludes
+var excs = map[string]reflect.Value{}
+
+// Init is used to intialize an executable module prior to entrypoint
+func (hfs *HFS) Init(api lib.APIClient) {
+	hfs.api = api
+	//hfs.mutex = &sync.Mutex{}
+	// hfs.queue = make(map[string]map[string]NMut)
```
Comment (kpelzel) [inline modules/hostfrequencyscaling/hostfrequencyscaling.go]: Remove these comments if they're not used

File: modules/hostfrequencyscaling/hostfrequencyscaling.go [hunk 2]
```diff
@@ -0,0 +1,516 @@
...
+		ScalingMinFreq:     scalingMinFreqq,
+		ScalingMaxFreq:     scalingMaxFreqq,
+		ScalingCurFreq:     scalingCurFreqq,
+
+		CPUCurFreq: cpuCurFreqq,
+		CPUMinFreq: cpuMinFreqq,
+		CPUMaxFreq: cpuMaxFreqq,
+	}
+	fmt.Println("Host:", currentScalingConfig.HostAddress, "Scaling Governor:", currentScalingConfig.CurScalingGovernor, "Scaling Current Freq:", currentScalingConfig.ScalingCurFreq, "CPU Current Freq:", currentScalingConfig.CPUCurFreq, "Scaling Min Freq:", currentScalingConfig.ScalingMinFreq, "Scaling Max Freq:", currentScalingConfig.ScalingMaxFreq, "CPU Min Freq:", currentScalingConfig.CPUMinFreq, "CPU Max Freq:", currentScalingConfig.CPUMaxFreq)
+	fmt.Println("")
```
Comment (kpelzel) [inline modules/hostfrequencyscaling/hostfrequencyscaling.go]: Use hfs.api.Logf for this and use '\n' if you need an extra newline

File: modules/hostthermaldiscovery/hostthermaldiscovery.go [hunk 3]
```diff
@@ -0,0 +1,273 @@
+/* HostThermalDiscovery.go: This module performs CPU Thermal discovery of HPC nodes through in-band (through OS) mechanism.
+ *
+ * Author: Ghazanfar Ali, ghazanfar.ali@ttu.edu; Kevin Pelzel <kevinpelzel22@gmail.com>; J. Lowell Wofford <lowell@lanl.gov>
+ *
+ * This software is open source software available under the BSD-3 license.
+ * Copyright (c) 2018, Triad National Security, LLC
+ * See LICENSE file for details.
+ */
+
+//go:generate protoc -I ../../core/proto/include -I proto --go_out=plugins=grpc:proto proto/pxe.proto
```
Comment (kpelzel) [inline modules/hostthermaldiscovery/hostthermaldiscovery.go]: remove this comment or replace with the correct file (not proto/pxe.proto)

File: core/StateSyncEngine.go [hunk 4]
```diff
@@ -433,6 +437,7 @@ func (sse *StateSyncEngine) nodeToMessage(to lib.NodeID, n lib.Node) (msg *pb.St
 	m := &pb.StateSyncMessage{}
 	m.Id = sse.self.Binary()
 	m.Message = n.Binary()
+	sse.Logf(lib.LLDEBUG, "converting node to binary: %v", string(n.JSON()))
```
Comment (kpelzel) [inline core/StateSyncEngine.go]: Remove debug line

File: kraken/main.go.tpl [hunk 5]
```diff
@@ -197,7 +197,10 @@ func main() {
 	// subscribe our listener
 	k.Ctx.SubChan <- sclist
 
-	k.Sme.Thaw()
+	// Thaw if full state
+	if len(parents) == 0 {
+		k.Sme.Thaw()
+	}
```
Comment (kpelzel) [inline kraken/main.go.tpl]: We might want this in a separate PR, but ask Lowell.

File: modules/hostfrequencyscaling/hostfrequencyscaling.go [hunk 6]
```diff
@@ -0,0 +1,516 @@
...
+	PxeURL string = "type.googleapis.com/proto.RPi3/Pxe"
+
+	// ModuleStateURL refers to module state
+	ModuleStateURL string = "/Services/hostfrequencyscaling/State"
+
+	// HostThermalStateURL points to Thermal extension
+	HostThermalStateURL string = "type.googleapis.com/proto.HostThermal/State"
+
+	// NodeIPURL provides node IP address
+	NodeIPURL string = "type.googleapis.com/proto.IPv4OverEthernet/Ifaces/0/Ip/Ip"
```
Comment (kpelzel) [inline modules/hostfrequencyscaling/hostfrequencyscaling.go]: These constant values _probably_ don't need to be exported, but it probably doesn't really matter. Ask Lowell

**Comment by kpelzel** [comment]
I just realized a lot of this PR contains changes from https://github.com/hpc/kraken/pull/112. We should make sure that gets merged first, then do a rebase.

File: config/pipxe.yaml [hunk 7]
```diff
@@ -14,10 +14,12 @@ targets:
...
-
+  - github.com/hpc/kraken/extensions/HostThermal
+  - github.com/hpc/kraken/extensions/RFAggregatorServer
 # included modules
 modules:
   - github.com/hpc/kraken/modules/restapi
   - github.com/hpc/kraken/modules/websocket
   - github.com/hpc/kraken/modules/rfpipower
   - github.com/hpc/kraken/modules/pipxe
+  - github.com/hpc/kraken/modules/rffrequencyscaling
```
[the remark this answers is not in the document]
Comment (ghazanfarttu) [reply config/pipxe.yaml]: rfAggregatorServer is merged as part of rfthermaldiscovery PR (closed). rffrequencyscaling is proposed under cpu frequency scaling PR (Open). Hostfrequencyscaling uses pipxe-child.yaml file

File: core/StateSyncEngine.go (hunk 4 above)
Comment (ghazanfarttu) [reply core/StateSyncEngine.go]: Fixed by commit 350d25b37ab43747487a4510feba5b17e9fce196

File: modules/hostthermaldiscovery/hostthermaldiscovery.go (hunk 3 above)
Comment (ghazanfarttu) [reply modules/hostthermaldiscovery/hostthermaldiscovery.go]: Fixed by commit 350d25b37ab43747487a4510feba5b17e9fce196

File: core/StateMutationEngine.go [hunk 8]
```diff
@@ -1340,6 +1342,7 @@ func (sme *StateMutationEngine) handleEvent(v lib.Event) {
 			sme.activeMutex.Unlock()
 		}
 	case StateChange_UPDATE:
+		sme.log.Logf(lib.LLDEBUG, "sme got a state change update event for %v %v %v", node, url, sce.Value)
```
[the remark this answers is not in the document]
Comment (ghazanfarttu) [reply core/StateMutationEngine.go]: Fixed by commit 350d25b37ab43747487a4510feba5b17e9fce196

File: core/StateDifferenceEngine.go [hunk 9]
```diff
@@ -430,6 +430,7 @@ func (n *StateDifferenceEngine) Run() {
 			n.Logf(DDEBUG, "processing discovery: si (%s) url (%s) id(%s)", data.ID, url, data.ValueID)
 			if !ok {
 				n.Logf(ERROR, "got discover, but can't lookup value: si (%s) url (%s) id(%s)", data.ID, url, data.ValueID)
+				n.Logf(ERROR, "registry: %+v", Registry.Discoverables[data.ID])
```
[the remark this answers is not in the document]
Comment (ghazanfarttu) [reply core/StateDifferenceEngine.go]: Fixed by commit 350d25b37ab43747487a4510feba5b17e9fce196

File: core/StateDifferenceEngine.go [hunk 10]
```diff
@@ -525,7 +526,7 @@ func (n *StateDifferenceEngine) bulkUpdateByType(dsc bool, ms []lib.Node) (r []l
 		utype = StateChange_CFG_UPDATE
 	}
 	for _, v := range diff {
-		evs = append(evs, NewStateChangeEvent(utype, v, reflect.Value{}))
+		evs = append(evs, NewStateChangeEvent(utype, v, reflect.ValueOf(r)))
```
[the remark this answers is not in the document]
Comment (ghazanfarttu) [reply core/StateDifferenceEngine.go]: Fixed by commit 350d25b37ab43747487a4510feba5b17e9fce196

File: core/StateMutationEngine.go [hunk 11]
```diff
@@ -914,6 +914,7 @@ func (sme *StateMutationEngine) startNewMutation(node string) {
 	defer p.mutex.Unlock()
 
 	sme.activeMutex.Lock()
+
```
[the remark this answers is not in the document]
Comment (ghazanfarttu) [reply core/StateMutationEngine.go]: Fixed by commit 350d25b37ab43747487a4510feba5b17e9fce196

File: modules/hostfrequencyscaling/hostfrequencyscaling.go (hunk 2 above)
Comment (ghazanfarttu) [reply modules/hostfrequencyscaling/hostfrequencyscaling.go]: Fixed by commit 350d25b37ab43747487a4510feba5b17e9fce196

File: core/Node.go [hunk 12]
```diff
@@ -399,9 +407,11 @@ func (n *Node) Diff(node lib.Node, prefix string) (r []string, e error) {
 		}
 	}
 	for _, u := range eleft {
-		r = append(r, lib.URLPush(prefix, u))
+		r = append(r, fmt.Sprintf("%s%s", prefix, u))
```
[the remark this answers is not in the document]
Comment (kpelzel) [reply core/Node.go]: Addressed in issue https://github.com/hpc/kraken/issues/142 so these changes can stay here

File: modules/hostfrequencyscaling/hostfrequencyscaling.go [hunk 13]
```diff
@@ -0,0 +1,677 @@
...
+	// 		case <-thermalCheckTick.C:
+	// 			if hfs.psEnforced == true {
+	// 				go hfs.CheckThermalThreshold()
+	// 			}
+
+	// 			break
+	// 		}
+	// 	}
+
+	// }
```
Comment (kpelzel) [inline modules/hostfrequencyscaling/hostfrequencyscaling.go]: remove commented code

File: modules/hostfrequencyscaling/hostfrequencyscaling.go [hunk 14]
```diff
@@ -0,0 +1,677 @@
...
+		// 	lib.Event_DISCOVERY,
+		// 	url,
+		// 	&core.DiscoveryEvent{
+		// 		URL:     url,
+		// 		ValueID: profileMap["performance"],
+		// 	},
+		// )
+		// hfs.dchan <- ev
+	}
+	//hfs.api.Logf(lib.LLERROR, "*** T E M P ***: %v", currentThermal/1000)
```
Comment (kpelzel) [inline modules/hostfrequencyscaling/hostfrequencyscaling.go]: removed commented code

File: modules/hostfrequencyscaling/hostfrequencyscaling.go [hunk 15]
```diff
@@ -0,0 +1,677 @@
...
+	// 	ev := core.NewEvent(
+	// 		lib.Event_DISCOVERY,
+	// 		url,
+	// 		&core.DiscoveryEvent{
+	// 			URL:     url,
+	// 			ValueID: currentScalingConfig.CurScalingGovernor,
+	// 		},
+	// 	)
+	// 	hfs.dchan <- ev
+	// }
```
Comment (kpelzel) [inline modules/hostfrequencyscaling/hostfrequencyscaling.go]: remove commented code

File: modules/hostthermaldiscovery/hostthermaldiscovery.go [hunk 16]
```diff
@@ -157,11 +185,73 @@ func (hostDisc *HostDisc) Entry() {
...
+}
+
+// DiscFreqScaler cpu frequency scaler
+func (hostDisc *HostDisc) DiscFreqScaler() {
+
+	hostFreqScaler := hostDisc.ReadFreqScaler()
+	// if hostFreqScaler == hostDisc.preFreqScaler {
+	// 	// no change in frequency scaler so no need to generate discovery event
+	// 	return
+	// }
```
Comment (kpelzel) [inline modules/hostthermaldiscovery/hostthermaldiscovery.go]: remove commented code

File: modules/hostthermaldiscovery/hostthermaldiscovery.go [hunk 17]
```diff
@@ -157,11 +185,73 @@ func (hostDisc *HostDisc) Entry() {
...
+	hostDisc.preFreqScaler = hostFreqScaler
+
+	vid := profileMap[hostFreqScaler]
+
+	//hostDisc.api.Logf(lib., "PRINTSCALER: %s", vid)
```
Comment (kpelzel) [inline modules/hostthermaldiscovery/hostthermaldiscovery.go]: remove commented code

File: modules/hostthermaldiscovery/hostthermaldiscovery.go [hunk 18]
```diff
@@ -157,11 +185,73 @@ func (hostDisc *HostDisc) Entry() {
...
+// ReadFreqScaler cpu frequency scaler
+func (hostDisc *HostDisc) ReadFreqScaler() string {
+
+	basePath := hostDisc.cfg.GetFreqSensorUrl()
+	bscalingGovernor, err := ioutil.ReadFile(basePath + "scaling_governor")
+	if err != nil {
+		hostDisc.api.Logf(lib.LLERROR, "Reading CPU thermal sensor failed: %v", err)
+		return ""
+	}
+	//fscalingGovernor := strings.TrimSuffix(string(bscalingGovernor), "\n")
```
Comment (kpelzel) [inline modules/hostthermaldiscovery/hostthermaldiscovery.go]: remove commented code

File: modules/hostthermaldiscovery/hostthermaldiscovery.go [hunk 19]
```diff
@@ -251,15 +341,24 @@ func (hostDisc *HostDisc) lambdaStateDiscovery(v CPUTempObj) (string, int32) {
...
 
-	lowerHigh := cpuThermalThresholds["CPUThermalThresholds"].GetLowerHigh()
-	upperHigh := cpuThermalThresholds["CPUThermalThresholds"].GetUpperHigh()
+	// lowerHigh := cpuThermalThresholds["CPUThermalThresholds"].GetLowerHigh()
+	// upperHigh := cpuThermalThresholds["CPUThermalThresholds"].GetUpperHigh()
 
-	lowerCritical := cpuThermalThresholds["CPUThermalThresholds"].GetLowerCritical()
-	upperCritical := cpuThermalThresholds["CPUThermalThresholds"].GetUpperCritical()
+	// lowerCritical := cpuThermalThresholds["CPUThermalThresholds"].GetLowerCritical()
+	// upperCritical := cpuThermalThresholds["CPUThermalThresholds"].GetUpperCritical()
```
Comment (kpelzel) [inline modules/hostthermaldiscovery/hostthermaldiscovery.go]: remove commented code

**Pull request merged**